### PR TITLE
state: replicationv2: state-machine: Implement state machine

### PR DESCRIPTION
**Pull request by joeykraut**
### Purpose
This PR implements the `RaftStateMachine` trait for the new consensus engine. Additionally this PR fixes a few bugs with storage that were discovered while integrating.

Not all tests pass, snapshotting needs to be implemented before the `openraft` suite passes

### Testing
- Non-state tests pass

### CODE REVIEW

File: state/src/storage/tx/mod.rs [hunk 1]
```diff
@@ -107,6 +107,7 @@ impl<'db> StateTxn<'db, RW> {
             MPC_PREPROCESSING_TABLE,
             NODE_METADATA_TABLE,
             RAFT_METADATA_TABLE,
+            RAFT_LOGS_TABLE,
```
Comment (akirillo) [inline state/src/storage/tx/mod.rs:110]: do we need to increment `NUM_TABLES`?

File: state/src/storage/tx/raft_log.rs [hunk 2]
```diff
@@ -233,13 +234,11 @@ impl<'db> StateTxn<'db, RW> {
...
-        let first_log = self.first_raft_log_index()?.unwrap_or(0);
-
-        // Purge the logs
         let mut log_cursor = self.logs_cursor()?;
+
+        let idx_lsn = lsn_to_key(index);
         log_cursor.seek(&idx_lsn)?;
-        for _ in first_log..=index {
+        while log_cursor.has_current()? {
             log_cursor.delete()?;
```
Comment (akirillo) [inline state/src/storage/tx/raft_log.rs:242]: just want to confirm that `log_cursor.delete()` will not progress the cursor's position, right?

**Review by akirillo — APPROVED** [review]

File: state/src/storage/tx/mod.rs (hunk 1 above)
Comment (joeykraut) [reply state/src/storage/tx/mod.rs:110]: no, this was already created elsewhere i added it here for automatic creation

File: state/src/storage/tx/raft_log.rs (hunk 2 above)
Comment (joeykraut) [reply state/src/storage/tx/raft_log.rs:242]: it _does_ move the cursor, this is the desired behavior and there is a test for this

**Pull request merged**